### PR TITLE
Generated names refactoring & support for function breakpoints on local functions

### PR DESCRIPTION
**Pull request by tmat**
Refactor GeneratedNames class - move parsing methods to GeneratedNameParser shared constants to GeneratedNameConstants.
This allows us to link GeneratedNameParser into FunctionResolver project, parse synthesized methods names and implement support for function breakpoints placement on local functions.

Fixes https://github.com/dotnet/roslyn/issues/55242

### CODE REVIEW

**Comment by tmat** [comment]
@cston @davidwengier PTAL

**Comment by davidwengier** [comment]
> Refactor GeneratedNames class

You always find a way to cause me conflicts 😛

**Review by davidwengier — APPROVED** [review]

File: src/Compilers/VisualBasic/Portable/Symbols/SynthesizedSymbols/GeneratedNameConstants.vb [hunk 1]
```diff
@@ -0,0 +1,57 @@
+﻿' Licensed to the .NET Foundation under one or more agreements.
+' The .NET Foundation licenses this file to you under the MIT license.
+' See the LICENSE file in the project root for more information.
+
+Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
+    Friend NotInheritable Class GeneratedNameConstants
+        Public Const DotReplacementInTypeNames As Char = "-"c
```
Comment (cston) [inline src/Compilers/VisualBasic/Portable/Symbols/SynthesizedSymbols/GeneratedNameConstants.vb:7]: >Public [](http://example.com/codeflow?start=8&length=6)

Consider using `Friend` for all of these constants, to minimize the differences with the previous iteration of `GeneratedNames.vb`, and also so the accessibility is clear from the fields directly.

File: src/ExpressionEvaluator/Core/Test/FunctionResolver/CSharpFunctionResolverTests.cs [hunk 2]
```diff
@@ -1313,6 +1468,38 @@ static async Task F()
...
+}
+";
+            var compilation = CreateCompilation(source);
+            using var process = new Process(new Module(compilation.EmitToArray()));
+
+            var resolver = Resolver.CSharpResolver;
+            Resolve(process, resolver, "G", "Program.<<Main>$>g__G|0_1()", "F.G()", "F.<G>g__G|0_0()");
+            Resolve(process, resolver, "Program.G", "Program.<<Main>$>g__G|0_1()");
+            Resolve(process, resolver, "F.G", "F.G()", "F.<G>g__G|0_0()");
+            Resolve(process, resolver, "A.G.G");
```
Comment (cston) [inline src/ExpressionEvaluator/Core/Test/FunctionResolver/CSharpFunctionResolverTests.cs:1500]: >A.G.G [](http://example.com/codeflow?start=40&length=5)

Consider testing `"F.G.G"` as well.

File: src/Compilers/CSharp/Portable/Emitter/EditAndContinue/PEDeltaAssemblyBuilder.cs [hunk 3]
```diff
@@ -123,6 +124,10 @@ private static EmitBaseline.MetadataSymbols GetOrCreateMetadataSymbols(EmitBasel
         // internal for testing
         internal static IReadOnlyDictionary<AnonymousTypeKey, AnonymousTypeValue> GetAnonymousTypeMapFromMetadata(MetadataReader reader, MetadataDecoder metadataDecoder)
         {
+            // In general, the anonymous type name is "<{module-id}>f__AnonymousType{index}#{submission-index}",
+            // but EnC is not supported for modules nor submissions. Hence we only look for type names with no module id and no submission index.
+            const string AnonymousNameWithoutModulePrefix = "<>f__AnonymousType";
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Emitter/EditAndContinue/PEDeltaAssemblyBuilder.cs:129]:
```suggestion
            internal const string AnonymousNameWithoutModulePrefix = "<>f__AnonymousType";
```

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/GeneratedNameParser.cs [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+            }
+
+            int localFunctionNameStart = closeBracketOffset + 2 + GeneratedNameConstants.SuffixSeparator.Length;
+            if (localFunctionNameStart >= generatedName.Length)
+            {
+                return false;
+            }
+
+            int localFunctionNameEnd = generatedName.IndexOf(GeneratedNameConstants.LocalFunctionNameTerminator, localFunctionNameStart);
+            if (localFunctionNameEnd < 0)
```
[the remark this answers is not in the document]
Comment (tmat) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/GeneratedNameParser.cs:130]: IndexOf docs: "Return: The zero-based index position of value from the start of the string if that character is found, or -1 if it is not."

File: src/ExpressionEvaluator/Core/Test/FunctionResolver/CSharpFunctionResolverTests.cs (hunk 2 above)
Comment (tmat) [reply src/ExpressionEvaluator/Core/Test/FunctionResolver/CSharpFunctionResolverTests.cs:1500]: `A` was meant to be `F` :)

**Review by jaredpar — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Emitter/EditAndContinue/PEDeltaAssemblyBuilder.cs (hunk 3 above)
Comment (tmat) [reply src/Compilers/CSharp/Portable/Emitter/EditAndContinue/PEDeltaAssemblyBuilder.cs:129]: This is local constant.

**Pull request merged**